NDArray Map

johnc1231 opened the pull request:
This PR:

- Introduces mapping over NDArrays in JVM emitter. 
- Introduces the `NDArrayEmitter` class, mimicking the cxx analogue. This class is used as the basis of our `NDArray` deforesting efforts (see `emitLoops`)
- Fixes a bug in `PStruct`'s load field function. 
- Introduces useful helper functions on `PNDArray`: `numElements` and `makeDefaultStrides`

Ready for review

catoverdrive commented on hail/src/main/scala/is/hail/expr/ir/Emit.scala:
```diff
@@ -1911,4 +1883,101 @@ private class Emit(
+        emitLoops(dataSrvb),
+        dataSrvb.end()
+      )
+    }
```
Don't need the enclosing `{}`

catoverdrive commented on hail/src/main/scala/is/hail/expr/types/physical/PNDArray.scala:
```diff
@@ -29,4 +30,94 @@ final case class PNDArray(elementType: PType, nDims: Int, override val required:
+      srvb.advance(),
+      srvb.addInt(offset),
+      srvb.advance(),
+      srvb.addIRIntermediate(this.representation.fieldType("shape").asInstanceOf[PTuple])(shape),
```
Do you need the `asInstanceOf[T]` casts here? `addIRIntermediate` should work without them.

johnc1231 commented on hail/src/main/scala/is/hail/expr/types/physical/PNDArray.scala:
```diff
@@ -29,4 +30,94 @@ final case class PNDArray(elementType: PType, nDims: Int, override val required:
+      0L
+    }
+    else {
+      val totalElements = mb.newField[Long]
```
Good idea

johnc1231 commented on hail/src/main/scala/is/hail/expr/types/physical/PNDArray.scala:
```diff
@@ -29,4 +30,94 @@ final case class PNDArray(elementType: PType, nDims: Int, override val required:
+    Code (
+      srvb.start(),
+      tupleStartAddress := srvb.offset,
+      // Fill with 0s, then backfill with actual data
```
Cool, didn't know I could use allocate like that

johnc1231 commented:
In the process of addressing your comments, I added this function: https://github.com/hail-is/hail/pull/6983/files#diff-544c40ab96249311692c1b9f8aec6fd8R21

Otherwise I couldn't get the Array.fold version of `numElements` to typecheck

catoverdrive commented on hail/src/main/scala/is/hail/expr/types/physical/PNDArray.scala:
```diff
@@ -29,4 +30,76 @@ final case class PNDArray(elementType: PType, nDims: Int, override val required:
+      0L
+    }
+    else {
+      Array.range(0, nDims).foldLeft(Code(1L)) { (prod, idx) => prod * getShapeAtIdx(idx) }
```
you can call the `const` function in the asm4s package for the Long -> Code[Long] coercion and you can also explicitly specify the type parameters to the `foldLeft` function.

johnc1231 replied:
Thanks, forgot about these options

catoverdrive commented on hail/src/main/scala/is/hail/asm4s/Code.scala:
```diff
@@ -18,6 +18,13 @@ object Code {
     }
   }
 
+  def apply[S1](c1: Code[S1]): Code[S1] =
```
This function should just return `c1` unchanged, but I think I might prefer if the typing issue was addressed in either of the other ways I mentioned. (up to you.)

johnc1231 replied:
Yeah, I'll just get rid of this and use const.
